File: app/src/main/java/com/github/doomsdayrs/apps/shosetsu/backend/settings/SettingsController.java
package com.github.doomsdayrs.apps.shosetsu.backend.settings;

import android.content.SharedPreferences;
import android.graphics.Color;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import com.github.doomsdayrs.apps.shosetsu.backend.database.Database;
import com.github.doomsdayrs.apps.shosetsu.variables.Settings;

/**
 * This file is part of Shosetsu.
 * Shosetsu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * Foobar is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * You should have received a copy of the GNU General Public License
 * along with Shosetsu.  If not, see https://www.gnu.org/licenses/ .
 * ====================================================================
 * Shosetsu
 * 14 / June / 2019
 *
 * @author github.com/doomsdayrs
 */
public class SettingsController {
    public static SharedPreferences download;
    public static SharedPreferences view;
    public static SharedPreferences advanced;
    public static SharedPreferences tracking;

    /**
     * Initializes the settings
     */
    public static void init() {
        Settings.ReaderTextColor = view.getInt("ReaderTextColor", Color.BLACK);
        Settings.ReaderTextBackgroundColor = view.getInt("ReaderBackgroundColor", Color.WHITE);
    }

    /**
     * Checks if online
     *
     * @return true if so, otherwise false
     */
    public static boolean isOnline() {
        NetworkInfo activeNetwork = Settings.connectivityManager.getActiveNetworkInfo();
        if (activeNetwork != null)
            return activeNetwork.getType() == ConnectivityManager.TYPE_WIFI || activeNetwork.getType() == ConnectivityManager.TYPE_MOBILE;
        return false;
    }

    /**
     * Is reader in night mode
     *
     * @return true if so, otherwise false
     */
    public static boolean isReaderLightMode() {
        return Settings.ReaderTextColor == Color.BLACK;
    }

    /**
     * Sets the reader color
     *
     * @param text       Color of text
     * @param background Color of background
     */
    private static void setReaderColor(int text, int background) {
        Settings.ReaderTextColor = text;
        Settings.ReaderTextBackgroundColor = background;
        view.edit()
                .putInt("ReaderTextColor", text)
                .putInt("ReaderBackgroundColor", background)
                .apply();
    }

    /**
     * Swaps the reader colors
     */
    public static void swapReaderColor() {
        if (isReaderLightMode())
            setReaderColor(Color.WHITE, Color.BLACK);
        else
            setReaderColor(Color.BLACK, Color.WHITE);
    }


    /**
     * Gets y position of a bookmark
     * TODO Add novelURL as data to search with
     *
     * @param chapterURL chapter chapterURL
     * @return y position
     */
    public static int getYBookmark(String chapterURL) {
        return Database.getY(chapterURL);
    }

    /**
     * Toggles bookmark
     *
     * @param chapterURL imageURL of chapter
     * @return true means added, false means removed
     */
    public static boolean toggleBookmarkChapter(String chapterURL) {

        if (Database.isBookMarked(chapterURL)) {
            Database.setBookMark(chapterURL, 0);
            return false;
        } else {
            Database.setBookMark(chapterURL, 1);
            return true;
        }
    }

    /**
     * Updates scroll position
     * TODO Novel chapterURL as well as parameter
     *
     * @param chapterURL chapterURL of chapter
     * @param y          y integer
     */
    public static void setScroll(String chapterURL, int y) {
        Database.updateY(chapterURL, y);
    }

    //Methods below when tracking system setup

    @SuppressWarnings({"EmptyMethod", "unused"})
    public static boolean isTrackingEnabled() {
        return tracking.getBoolean("enabled", false);
    }

    @SuppressWarnings({"EmptyMethod", "unused"})
    public static void addTracker() {
    }
}

File: app/src/main/java/app/shosetsu/android/ui/settings/SettingsController.kt
package app.shosetsu.android.ui.settings

import android.view.View
import app.shosetsu.android.common.SettingsCard
import app.shosetsu.android.ui.settings.SettingsController.Types.*
import app.shosetsu.android.ui.settings.adapter.SettingsAdapter
import app.shosetsu.android.view.base.RecyclerController.BasicRecyclerController
import com.github.doomsdayrs.apps.shosetsu.R

/*
 * This file is part of Shosetsu.
 *
 * Shosetsu is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Shosetsu is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Shosetsu.  If not, see <https://www.gnu.org/licenses/>.
 */

/**
 * Shosetsu
 * 9 / June / 2019
 */
class SettingsController : BasicRecyclerController<SettingsAdapter, SettingsCard>() {
	override val viewTitleRes: Int = R.string.settings
	override var recyclerArray: ArrayList<SettingsCard>
		get() = arrayListOf(
				SettingsCard(DOWNLOAD),
				SettingsCard(UPDATE),
				SettingsCard(VIEW),
				SettingsCard(READER),
				SettingsCard(ADVANCED),
				SettingsCard(INFO),
				SettingsCard(BACKUP)
		)
		set(_) {}


	override fun onViewCreated(view: View) {}

	override fun difAreItemsTheSame(oldItem: SettingsCard, newItem: SettingsCard): Boolean =
			oldItem.id == newItem.id

	override fun setupRecyclerView() {
		super.setupRecyclerView()
		recyclerView.setHasFixedSize(true)
	}

	override fun createRecyclerAdapter(): SettingsAdapter =
			SettingsAdapter(recyclerArray, router)

	enum class Types {
		DOWNLOAD, // Settings for download options
		UPDATE,
		VIEW, // Settings for application appearance
		ADVANCED, // Settings that control more advanced application features
		INFO,  // Information of the app
		BACKUP, // Settings for backup and restoring data
		READER // Settings for reading novels in application
	}
}